algorithms/src/std_algorithms/impl/Kokkos_Constraints.hpp

```diff
@@ -215,6 +215,28 @@ KOKKOS_INLINE_FUNCTION void expect_valid_range(IteratorType first,
   (void)last;
 }
 
+//
+// Check if iterators are overlapped
+//
+template <typename IteratorType1, typename IteratorType2>
+KOKKOS_INLINE_FUNCTION void expect_no_overlap(IteratorType1 first,
+                                              IteratorType1 last,
+                                              IteratorType2 s_first,
+                                              IteratorType2 s_last) {
+  if constexpr (std::is_constructible_v<IteratorType2, IteratorType1>) {
+    IteratorType1 s_first1(s_first), s_last1(s_last);
+    KOKKOS_EXPECTS(first > s_last1 || last < s_last1);
```

Comment:
```suggestion
    KOKKOS_EXPECTS(first > s_last1 || last < s_first1);
```

Reply:
I was imagining a scenario where you could have iterators of the same type with the same stride but different starting pointers so that they don't overlap. For example, the location in memory would be alternating between the first iterator and the second iterator.

Reply:
I would only try to detect overlapping memory for 
- View types with stride 1
- raw pointers
- std::contiguous_iterators

Reply:
I think you can check for all `Kokkos::view::iterator` in case the referenced view is the same, even with non unit stride. This would case the most common case, wouldn't it?

Reply:
We need to add a method to `Kokkos::iterator` to compare internal views?

Reply:
That is nice. So the check would be something like this? (assuming the View access operator returns a reference)

```C++
auto next_first = first;
auto stride  = &*(++next_first) - &*first;
if constexpr (std::is_constructible_v<IteratorType2, IteratorType1>) {
  IteratorType1 s_first1(s_first), s_last1(s_last);
  auto first_diff = &*(first) - &*(s_first1);
  bool is_no_overlap = first_diff % stride;
  KOKKOS_EXPECTS(first > s_last1 || last < s_first1 || is_no_overlap);
}
```